Source/WebCore/dom/Element.cpp

```diff
@@ -2298,19 +2298,31 @@ void Element::setElementsArrayAttribute(const QualifiedName& attributeName, std:
     }
 }
 
-void Element::classAttributeChanged(const AtomString& newClassString)
+void Element::classAttributeChanged(const AtomString& newClassString, AttributeModificationReason reason)
 {
     // Note: We'll need ElementData, but it doesn't have to be UniqueElementData.
     if (!elementData())
         ensureUniqueElementData();
 
-    {
+    ([&]() ALWAYS_INLINE_LAMBDA {
+        if (reason == AttributeModificationReason::Parser) {
+            // If ElementData is ShareableElementData created in parserSetAttributes,
+            // it is possible that SpaceSplitString is already created and set.
+            // We also do not need to invalidate caches / styles since it is not inserted to the tree yet.
+            if (elementData()->classNames().keyString() == newClassString)
+                return;
+            auto shouldFoldCase = document().inQuirksMode() ? SpaceSplitString::ShouldFoldCase::Yes : SpaceSplitString::ShouldFoldCase::No;
+            SpaceSplitString newClassNames(newClassString, shouldFoldCase);
+            elementData()->setClassNames(WTFMove(newClassNames));
+            return;
+        }
+
         auto shouldFoldCase = document().inQuirksMode() ? SpaceSplitString::ShouldFoldCase::Yes : SpaceSplitString::ShouldFoldCase::No;
         SpaceSplitString newClassNames(newClassString, shouldFoldCase);
         Style::ClassChangeInvalidation styleInvalidation(*this, elementData()->classNames(), newClassNames);
         document().invalidateQuerySelectorAllResultsForClassAttributeChange(*this, elementData()->classNames(), newClassNames);
         elementData()->setClassNames(WTFMove(newClassNames));
-    }
+    }());
 
     if (hasRareData()) {
```

Comment:
Maybe we should re-order so that we do this invalidation of classList first.
That way, we don't need to wrap the code in a lambda. WDYT?

Reply:
Sounds good! Changed.